staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go

```diff
@@ -338,6 +351,27 @@ func New(opts Options) (*Authenticator, error) {
 	return authenticator, nil
 }
 
+// roundTripper is a http.RoundTripper that rewrites the issuer URL to the
+// discovery URL for the /.well-known/openid-configuration endpoint.
+type roundTripper struct {
+	rt           http.RoundTripper
+	discoveryURL *url.URL
+}
+
+func (t *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	// If the discovery URL is set in authentication configuration, we should
+	// use the discovery URL to fetch the openid configuration. This is useful
+	// for self-hosted providers, for example, providers that run on top of
+	// Kubernetes itself.
+	if t.discoveryURL != nil {
+		if req.Method == http.MethodGet && strings.HasSuffix(req.URL.Path, "/.well-known/openid-configuration") {
+			req.URL.Host = t.discoveryURL.Host
+			req.URL.Scheme = t.discoveryURL.Scheme
+		}
+	}
+	return t.rt.RoundTrip(req)
+}
```

Comment:
This isn't correct for cases where the discovery and/or the issuer URL have a path component that is different.

i.e.

issuer URL = `https://hello.com/a/b/foo`
discovery URL = `https://bye.com/c/d/bar`

You will see a request for

`https://hello.com/a/b/foo/.well-known/openid-configuration`

and rewrite it to

`https://bye.com/a/b/foo/.well-known/openid-configuration`

instead of the correct URL

`https://bye.com/c/d/bar/.well-known/openid-configuration`